Taxonomy model for Label is not correct

jncosideout opened the pull request:
## PR Description

While discussing how to create a mock-up for [Issue #707 ](https://github.com/openfoodfacts/openfoodfacts-ios/issues/707) @teolemon realized that Labels have not been implemented yet. I have taken the first steps in adding the necessary methods to TaxonomyService, TaxonomyParser PersistanceManager by mimicking the other Taxonomy methods in the protocol/API. 

I am stuck on how to implement the Label model. I copied the Category model because from my limited knowledge, they seem to have similar data models, and I noticed that the model for Mineral may be identical to Category. For clarity, I am talking about the models under Sources/Models/API/Taxonomies/

Type of Changes 

- [x] Fixes Issue #719 

## Checklist
 
Make sure you've done all the following (_Put an `x` in the boxes that apply._)
 
 - [x] If you have multiple commits please combine them into one commit by [squashing](https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit) them. 
 - [ ] Code is well documented
 - [ ] Included unit tests for new functionality
 - [ ] All user-visible strings are made translatable
 - [ ] Code passes Travis builds in your branch

jncosideout committed:
Add Labels to TaxonomiesService and Product view

Labels were partially implemented but still needed to be incorporated
into the API Models and Network Services. In doing so, it was discovered
that the Taxonomies weren't being downloaded to Realm on first launch of
the app as intended. After fixing that and adding the Label taxonomy to
Models, they were successfully retrieved and stored. Then, to allow
Labels to be displayed by ProductDetailViewController, they were
included in the OFFJson mapping. The downloaded taxonomies were then
used to translate Labels and code was added to fix translation of
Categories. Lastly, the functionality to link link Labels to their OFF
urls was added.

Closes #719

jncosideout commented on Sources/Network/TaxonomiesService.swift:
```diff
@@ -356,7 +357,28 @@ class TaxonomiesService: TaxonomiesApi {
+    
+    fileprivate func refreshLabels(_ callback: @escaping (_: Bool) -> Void) {
+        do {
+            let request = try TaxonomiesRequest(route: .getLabels, requestType: .get).asURLRequest()
```
Every TaxonomyRequest in each 'refresh' method passes .get to this method. That makes each one including this one throw an error because TaxonomiesRequest.asURLRequest() formats the path for only two matching cases:
```
private var path: String {
        switch (requestType, route) {
        case (.get, .getIngredientsAnalysisConfig):
            return route.rawValue + "/data/" + Endpoint.get
        case (.post, _):
            return Endpoint.post + route.rawValue
        default:
            return ""
        }
    }
```
So this is making it hard for me to test my code, I can't fetch the Labels. Any suggestions? What else should I know about this process?

aleene replied:
I do not get what you are saying. You have another case that you want to match?

jncosideout replied:
Yes, I would add 
`case (.get, _):`
to `var path`

but I only noticed that this is a problem only when the app first loads, because that is when `RootViewController` calls `taxonomiesApi.refreshTaxonomiesFromServerIfNeeded()`
At that point, none of the Taxonomies are loaded, including the new one for `Label` which I am trying to add in this PR.

So I want to give it the benefit of the doubt, that a new case _doesn't_ need to be added to `path`, because obviously the other taxonomies are working somehow, I just don't know how because I'm new to this project.

aleene replied:
I once added a new taxonomy, but I do not quite remember what I did, alas.

I do not see why you would want an extra case, as you always want to know the configuration parameters.

I wonder whether you need to up the model number for Realm, as you add new elements to Realm.

aleene replied:
That is in AppDelegate.swift.

aleene replied:
Or are the taxonomies not handled by Realm?

jncosideout replied:
During the refresh of Taxonomies, they appear to be saved with the PersistenceManager to Realm after being fetched from the server.

Again, I'm only going off what I've seen when debugging as soon as the app launches, so I'm sure I don't get the big picture yet. So adding an extra case appears to be the wrong thing to do.

Can you please explain to me why the Taxonomies are downloaded? I assume they aren't connected to products until a product is scanned, so it must be that at that point, a scanned product contains it's Taxonomy data, so why are the Taxonomies themselves (and all their possible values?) stored on the device?

aleene replied:
The taxonomies themselves are downloaded on first launch and regularly afterwards. I guess the main reason at the moment is to have an offline app. But I do not have a total insight.
By the way, I assume that you need to create a Label object to store the label data you retrieve. You can use another taxonomy as template (such as categories). However there might be some extra fields in a retrieved label object that must be added. You can check label.json to see what is sent/retrieved.

jncosideout replied:
Thanks Arnaud. Yes, that makes sense, I suspected it was for offline functionality. 
Also, I have already created a Label object to store the data, and I just copied the Category taxonomy. I will check Label.json to fill in any missing fields.

jncosideout replied:
Please see my comments on `TaxonomiesRequest` below where I explain why I made the changes discussed in this thread. I believe that none of the Taxonomies were being downloaded prior to making this change.

jncosideout commented on Sources/Network/TaxonomiesRequest.swift:
```diff
@@ -30,6 +30,8 @@ struct TaxonomiesRequest: URLRequestConvertible {
             return route.rawValue + "/data/" + Endpoint.get
         case (.post, _):
             return Endpoint.post + route.rawValue
+        case (.get, _):
+            return Endpoint.get + "/data/" + route.rawValue
```
Adding this case was the only way to allow the creation of URLrequests to succeed when `TaxonomyRequest.asURLRequest()` is called by any of the `refresh`{Taxonomy}`()` methods in `TaxonomiesService.downloadTaxonomies()`
Without that change, each of the taxonomy requests (except `refreshIngredientsAnalysisConfig()`) **always** fail.

The return value for the first case `(.get, .getIngredientsAnalysisConfig):`  never gets used by `refreshIngredientsAnalysisConfig()`. Instead, it uses `FilesRouter` to create its URL request. So it seems like that case should have been removed.

I also had to reorder parts of the return value for the case that I added, instead of  
`return route.rawValue + "/data/" + Endpoint.get`
I did
`return Endpoint.get + "/data/" + route.rawValue`
because that is how the actual URLs are formatted for the JSON data.

jncosideout replied:
@aleene, this is the change I was discussing with you last week. I believe that **none** of the Taxonomies were being downloaded prior to making this change. I know that you said: "I do not see why you would want an extra case, as you always want to know the configuration parameters," but cannot see what harm this change causes, and I really think it was necessary for downloading all the Taxonomies, including 'Label,' which I just added.

jncosideout replied:
I am 99% sure these changes were necessary because I believe that none of the Taxonomies were being downloaded prior to making this change.

jncosideout commented on Sources/Network/TaxonomiesService.swift:
```diff
@@ -20,7 +20,7 @@ enum TaxonomiesRoute: String {
     case getMinerals = "taxonomies/minerals.json"
     case getNucleotides = "taxonomies/nucleotides.json"
-    case getInvalidBarcodes = "invalid-barcodes.json"
+    case getInvalidBarcodes = "taxonomies/invalid-barcodes.json"
```
I made this change because the `refreshInvalidBarcodes()` request was always failing. But this url does not lead to the invalid-barcodes.json and I'm wondering if it has been removed entirely, I just can't find it.

teolemon replied:
The right url is https://static.openfoodfacts.org/data/invalid-barcodes.json

teolemon replied:
it's in a different directory than taxonomies

jncosideout commented on Sources/Models/API/Taxonomies/Label.swift:
```diff
@@ -0,0 +1,49 @@
+
+class Label: Object {
+
+    @objc dynamic var code = ""
+
+    let parents = List<String>()
+    let children = List<String>()
+    let names = List<Tag>()
```
After making the changes in my second commit ("Allow TaxonomyRequests to succeed when creating URLs..." 210e186), the data for labels was successfully saved to Realm (debug output: `Saved 2364 labels in taxonomy database.`) However, I just copied this model from Categories, and I noticed that almost all Labels and Categories have 'parents,' 'children,' and 'names' as attributes, some of them have other attrbs. like 'country,' 'eu-groups,' 'auth-url,' to name a few. Not sure how to handle all of them or if this is enough.

teolemon replied:
We probably don't use any of the attributes, in any of the taxonomies on iOS right now. If it's not too much extra work storing them, that would be useful, otherwise don't bother.

jncosideout replied:
Ok. The problem is, I don't know which attributes are common to all Labels, other than `parents, children, names.` I was trying to reverse-engineer the data model by picking random Labels in the database and looking at their attributes.

teolemon replied:
It's constantly evolving and attributes are not present on all labels.

jncosideout commented:
By adding `labelsTags` to the Product API model and mapping it to the OFFJson, Labels are now displayed on scan. However, they aren't translated correctly yet, despite having fixed the downloading of Taxonomies on launch of the app. 

As you can see below in the before and after pictures, the Labels are now included in the tableview of the ProductDetail Rows but they are not being translated into English, which is the default language on my testing device. 

The labels 'Bio, Bio européen, Agriculture UE/Non-UE' are the French translations of those Labels, and in English they should say 'en:organic, en:eu-organic, en:ab-agriculture-biologique, en:eu-non-eu-agriculture.'

> ### **Before**
![before_labels](https://user-images.githubusercontent.com/25991237/92145062-45a1ac00-eddd-11ea-8ea7-be4a5de16f17.jpg)

> ### **After**
![after_labels](https://user-images.githubusercontent.com/25991237/92145116-5d793000-eddd-11ea-8a00-dce4f1f9c199.jpg)

aleene commented:
Well, it looks good. Is now the translation the only remaining issue?

jncosideout commented:
@aleene 
Yes, still need to figure out translations. I don't know how the downloaded taxonomies are supposed to be used to to translate. Like I said above, I believe that the downloading of taxonomies was not working before I made my changes. So there must be some way to look up a Label in the downloaded taxonomy database, e.g. after scanning a product which has the 'Organic' Label to get its translations.

jncosideout commented:
> As you can see below in the before and after pictures, the Labels are now included in the tableview of the ProductDetail Rows but they are not being translated into English, which is the default language on my testing device.

@teolemon @aleene 
I have fixed the translation of the Labels. I started by checking how the Categories are translated, but I found that the code to do so was not working correctly. They were always being displayed in English, even when I changed the language setting to Français (see below):

> ### **Before**
Language : French
Categories : English 
Labels : English (displaying Labels_Tags values for testing)
![OrangeAmere-in_French-Labels_tags](https://user-images.githubusercontent.com/25991237/94351181-ea4d8d00-001b-11eb-8fee-439545ab1723.png)


> ### **After**
Language : French
Categories : French 
Labels : French
![OrangeAmere-in_French-Categories-translated](https://user-images.githubusercontent.com/25991237/94351162-cbe79180-001b-11eb-9dfb-09811f05a181.png)


Language : English 
Categories : English 
Labels : English
![OrangeAmere-in_English-Labels_translated](https://user-images.githubusercontent.com/25991237/94351176-ddc93480-001b-11eb-885e-be635d1bae9e.png)

[EDIT: Below are my findings from **before** I figured out how to use taxonomies to translate Labels and Categories]

The product I tested with was Orange Amère - Barcode: 3760036843626(EAN / EAN-13) 
The reason the Labels appeared in French and not English (which is the default language on my testing device) was because this particular product has different values for the ‘Labels’ on the dev server than on the production server.

On the [dev server](https://world.openfoodfacts.dev/api/v0/product/3760036843626.json), this product’s
‘label’ values are in French,
‘label_tags’ values are in English.
label_lc: fr

On the [production server](https://world.openfoodfacts.org/api/v0/product/3760036843626.json), this product’s
‘label’ values are in English,
‘label_tags’ values are in English.
label_lc: en

teolemon commented:
Let's focus on what production does. label_tags are always in English, and labels will be in the main language of the product. 80-90% of the values will be in label_tags in english, and the non-taxonomized values will also be there, with a language prefix.
Let's display what's already taxonomized, if possible in the user language for the phone, or in the main language of the product, or in English at the last resort.

jncosideout commented:
> label_tags are always in English, and labels will be in the main language of the product.

@teolemon 
Sorry for the confusion. Yes at first I used Labels_tags as a work-around to translate into English , then I realized they are always in English but they have the "en:" prefix and therefore that was not the correct way to do it.

If you look at my 2nd and 3rd screenshots in my last post you will see that the Labels are displaying their translated values, and if you look at the changes I made to `ProductDetailViewController` you will see I am using the downloaded taxonomies to fetch the translated names.

jncosideout commented on Sources/ViewControllers/Products/Detail/ProductDetailViewController.swift:
```diff
@@ -211,14 +211,22 @@ class ProductDetailViewController: ButtonBarPagerTabStripViewController, DataMan
-        createFormRow(with: &rows, item: product.labels, label: InfoRowKey.labels.localizedString)
+        createFormRow(with: &rows, item: product.labelsTags?.map({ (labelTag: String) -> NSAttributedString in
+            if let label = dataManager.label(forTag: labelTag) {
+                if let name = Tag.choose(inTags: Array(label.names), forLanguageCode: Bundle.main.currentLocalization) {
```
@teolemon 
Here is where I added the code to translate Labels by fetching the corresponding taxonomy of the Label.

jncosideout commented on Sources/ViewControllers/Products/Detail/ProductDetailViewController.swift:
```diff
@@ -211,14 +211,22 @@ class ProductDetailViewController: ButtonBarPagerTabStripViewController, DataMan
         createFormRow(with: &rows, item: product.categoriesTags?.map({ (categoryTag: String) -> NSAttributedString in
             if let category = dataManager.category(forTag: categoryTag) {
-                if let name = Tag.choose(inTags: Array(category.names)) {
+                if let name = Tag.choose(inTags: Array(category.names), forLanguageCode: Bundle.main.currentLocalization) {
```
@teolemon 
I realized that Categories weren't being translated either, so I added `currentLocalization` from the Bundle. This works better than `preferredLocalization` because `currentLocalization` gets updated when the user changes their language settings.

philippeauriach replied:
Instead of changing in every `Tag.choose` call, it would be better to replace the call to `preferredLocalizations.first` in the Tag.choose implementation, so that we have the same implementation and logic every where

jncosideout replied:
I agree. Plus, `currentLocalization` falls back to `String(Bundle.main.preferredLocalizations.first ?? "en")` so this is a safe change.

jncosideout commented on Sources/Helpers/OFFUrlsHelper.swift:
```diff
@@ -10,7 +10,7 @@ import UIKit
 class OFFUrlsHelper: NSObject {
 
-    static let codeLang = Bundle.main.preferredLocalizations.first ?? "en"
+    static let codeLang = Bundle.main.currentLocalization
```
@teolemon 
`preferredLocalizations` does not get updated dynamically. `currentLocalization` gets updated as soon as the user selects a new language in settings. 

I found that if I changed this from a `static let` to a static method the user does not need to restart the app in order to click on the links from the Categories and Labels on a product detail view to see the OFF webpage.

However, making this change meant changing one call to `OFFUrlsHelper.baseUrl` in IngredientsAnalysisView.swift, so I did not include that change in this commit, and I am experimenting with it to see if that change will cause any errors.

If it works however, I think it warrants its own PR. It would allow these links to point to the OFF urls subdomain of the user's language as soon as he switches languages.

jncosideout replied:
This has been addressed in #747 , which was just approved, and my latest commit has merged that PR in.

teolemon replied:
and we have also some other incorrect behaviours related to the same cause.
Possibly: https://github.com/openfoodfacts/openfoodfacts-ios/issues/684
https://github.com/openfoodfacts/openfoodfacts-ios/issues/662

aleene replied:
I do not remember it any more, but is the preferredLocalizations not the list that is used by IOS itself (in the preferences). It should be the default behaviour. The language changer is more an advanced feature for demonstration.

jncosideout replied:
@aleene 
 Correct, `preferredLocailzations` is the language of the iOS device settings. However, this change has been merged recently in PR #747 , so I think that is the direction we're taking. At least `currentLocailzation` defaults to `preferredLocailzations` if it hasn't been set, and if that fails too, it defaults to "en"
@teolemon and @philippeauriach can correct me if I'm wrong.

aleene replied:
OK, I am a bit at the loss why it is not working (anymore?). But maybe forgot to test some scenarios.

Not sure where, but I saw a discussion on changing the url. Beware that the prefix might limit a search to products sold in a particular country (have to check whether this is the case).

jncosideout replied:
> OK, I am a bit at the loss why it is not working (anymore?). But maybe forgot to test some scenarios.

It was working, in that the in-app webview was opening. The scenario this addresses is only for the hyperlinks to OFF.org, when the user changes the language in the app.

> `preferredLocalizations` ... should be the default behaviour.

It is the default on the app's first launch. But I think I understand your concern now; are you saying the app's language should change when the device's language changes and then again after the user changes it in the app? I believe that was not the case before I edited this code, instead the app's language stays the same even if the user changes the device's language. 

Of course, I could be wrong. I would like to know if the we are using `preferredLocalizations` or `currentLocalization` for the UI strings throughout the app. I'm guessing it's `currentLocalization` because that seems to be changed dynamically when the user changes languages in the app's settings. Please correct me if I'm wrong.

> Not sure where, but I saw a discussion on changing the url. Beware that the prefix might limit a search to products sold in a particular country (have to check whether this is the case).

I don't think this edit affects product search, only the hyperlinks to OFF.org. Are you saying that the OFF website is going to change the url address from "world-(lang).off.org" to just "(lang).off.org" or something like that?

jncosideout commented on Sources/Helpers/OFFUrlsHelper.swift:
```diff
@@ -46,6 +46,10 @@ class OFFUrlsHelper: NSObject {
+    }
+
     // Not sure if there is a taxonomy for this
     /*
```
Merged with the latest changes from this PR [Open weblinks to OFF urls in the language chosen in app settings, dynamically #747](https://github.com/openfoodfacts/openfoodfacts-ios/pull/747)

aleene reviewed (dismissed):
Many changes, so if it works for you, we can do some testing on develop.

jncosideout committed:
Merge branch 'develop' into Add-Labels-to-TaxonomiesService
